### ISSUE TITLE
LXD 5.21.1 daemon not starting due to 'zpool is missing' error

### ISSUE THREAD

**Issue by amcduffee** [canonical/lxd#13319]
# Required information

 * Distribution: Ubuntu
 * Distribution version: 18.04.6 LTS
 * The output of "snap list --all lxd core20 core22 core24 snapd":
 ```
# snap list --all lxd core20 core22 core24 snapd
Name    Version         Rev    Tracking       Publisher   Notes
core20  20240111        2182   latest/stable  canonical✓  base,disabled
core20  20240227        2264   latest/stable  canonical✓  base
core22  20231123        1033   latest/stable  canonical✓  base,disabled
core22  20240111        1122   latest/stable  canonical✓  base
lxd     5.20-f3dd836    27049  latest/stable  canonical✓  disabled
lxd     5.21.1-43998c6  28155  latest/stable  canonical✓  -
 ```
 * The output of "lxc info" or if that fails:
   * Kernel version: 5.4.0-150-generic
   * LXC version: 5.21.1
   * LXD version: 5.21.1
   * Storage backend in use: zfs

# Issue description
Some change in the recent snap refresh causes the `snap.lxd.daemon.service` to fail to start repeatedly.

The first indication of this was the following:
```
$ lxc ls
Error: LXD unix socket "/var/snap/lxd/common/lxd-user/unix.socket" not accessible: Get "http://unix.socket/1.0": dial unix /var/snap/lxd/common/lxd-user/unix.socket: connect: connection refused
```

It looks like the error occurred this morning after a snap refresh:
```
# snap logs lxd
2024-04-11T05:36:19-07:00 lxd.daemon[4709]: time="2024-04-11T05:36:19-07:00" level=debug msg="Initializing database gateway"
2024-04-11T05:36:19-07:00 lxd.daemon[4709]: time="2024-04-11T05:36:19-07:00" level=info msg="Starting database node" id=1 local=1 role=voter
2024-04-11T05:36:20-07:00 systemd[1]: snap.lxd.daemon.service: Main process exited, code=exited, status=1/FAILURE
2024-04-11T05:36:20-07:00 systemd[1]: snap.lxd.daemon.service: Failed with result 'exit-code'.
2024-04-11T05:36:20-07:00 systemd[1]: snap.lxd.daemon.service: Service hold-off time over, scheduling restart.
2024-04-11T05:36:20-07:00 systemd[1]: snap.lxd.daemon.service: Scheduled restart job, restart counter is at 6.
2024-04-11T05:36:20-07:00 systemd[1]: Stopped Service for snap application lxd.daemon.
2024-04-11T05:36:20-07:00 systemd[1]: snap.lxd.daemon.service: Start request repeated too quickly.
2024-04-11T05:36:20-07:00 systemd[1]: snap.lxd.daemon.service: Failed with result 'exit-code'.
2024-04-11T05:36:20-07:00 systemd[1]: Failed to start Service for snap application lxd.daemon.

# snap info lxd
name:      lxd
...
snap-id:      J60k4JY0HppjwOjW8dZdYc8obXKxujRu
tracking:     latest/stable
refresh-date: today at 05:36 PDT
...
installed:          5.21.1-43998c6            (28155) 107MB -
```

Checking the LXD server log shows the likely culprit:
```
# grep level=error /var/snap/lxd/common/lxd/logs/lxd.log
time="2024-04-11T10:59:30-07:00" level=error msg="Failed loading storage pool" err="Required tool 'zpool' is missing" pool=zfs
time="2024-04-11T10:59:30-07:00" level=error msg="Failed to start the daemon" err="Failed applying patch \"storage_move_custom_iso_block_volumes_v2\": Failed loading pool \"zfs\": Required tool 'zpool' is missing"
```

ZFS is definitely still installed because I can use the 'zpool' command without issue:
```
# zpool list
NAME    SIZE  ALLOC   FREE  EXPANDSZ   FRAG    CAP  DEDUP  HEALTH  ALTROOT
pool1  10.9T  4.39T  6.52T         -     2%    40%  1.00x  ONLINE  -
```

**Comment by tomponline** [comment]
LXD 5.21 requires ZFS 2.1 or newer, see release notes:

https://discourse.ubuntu.com/t/lxd-5-21-0-lts-has-been-released/42476

You only need to get the kernel ZFS version increased, not the OS userspace software, so you may find that using an Ubuntu HWE kernel can help, see discussion about this on the forum:

https://discourse.ubuntu.com/t/lxd-5-21-1-lts-has-been-released/43823/19

It should be noted that if you use the `latest/stable` snap track you will be on a rolling LXD release, and the minimum system requirements do change over time.

The alternative is to follow on of the LTS series, e.g. 5.0/stable or 5.21/stable as these will not have their minimum system requirements change during their supported life time.

See https://documentation.ubuntu.com/lxd/en/latest/installing/#choose-your-release

**Comment by tomponline** [comment]
I've asked and 18.04's HWE kernel doesn't support ZFS 2.1 either.

Do you have recent backups that you can restore from?

I would suggest switching to the LXD 5.0/stable LTS channel, but because you've now upgraded to LXD 5.21.1 the DB schema changes will mean its incompatible with LXD 5.0.x

**Comment by amcduffee** [comment]
I don't have backups to restore from, not for the LXD snap folder.  It is my development machine, so at least not production.

Fortunately our production cluster doesn't use ZFS or this would have been a major catastrophe.  All 5 machines in the cluster are tracking `latest/stable` as the installation was done years ago when that was the default behavior for `snap install lxd`.

Was there forward warning that this breaking change was going to occur?

Should I switch the production cluster to track `5.21/stable` since they are all at `5.21.1-43998c6` already?

**Comment by tomponline** [comment]
> Should I switch the production cluster to track `5.21/stable` since they are all at `5.21.1-43998c6` already?

Yes you should do that to ensure they are not on a rolling release.

**Comment by tomponline** [comment]
> I don't have backups to restore from, not for the LXD snap folder. It is my development machine, so at least not production.
> 
> Fortunately our production cluster doesn't use ZFS or this would have been a major catastrophe. All 5 machines in the cluster are tracking `latest/stable` as the installation was done years ago when that was the default behavior for `snap install lxd`.
> 
> Was there forward warning that this breaking change was going to occur?
> 
> Should I switch the production cluster to track `5.21/stable` since they are all at `5.21.1-43998c6` already?

Yes it was in the release notes about the change in requirement and the release was in the candidate channel for several weeks.

**Comment by amcduffee** [comment]
> > Should I switch the production cluster to track `5.21/stable` since they are all at `5.21.1-43998c6` already?
> 
> Yes you should do that to ensure they are not on a rolling release.

Ok, so a `snap refresh --channel=5.21/stable lxd` is all I need?  Also, it looks like that will revert from `5.21.1-43998c6` to `5.21.1-3f3c2a9`, is this likely to cause any issues that you are aware of?

**Comment by tomponline** [comment]
> I don't have backups to restore from, not for the LXD snap folder. It is my development machine, so at least not production.

I'm thinking one way out of this is to switch to 5.0/stable channel, delete the lxd database and do a `lxd recover ` to scan the pool for instances and rebuild the db.

I'll try this and let you know if works.

**Comment by tomponline** [comment]
> Ok, so a `snap refresh --channel=5.21/stable lxd` is all I need? Also, it looks like that will revert from `5.21.1-43998c6` to `5.21.1-3f3c2a9`, is this likely to cause any issues that you are aware of?

That's fine. They are built from different packaging branches but are equivalent.

**Comment by tomponline** [comment]
If your using a cluster also add `--cohort=”+”` see 

https://documentation.ubuntu.com/lxd/en/latest/howto/snap/#keep-cluster-members-in-sync

**Comment by simondeziel** [comment]
@amcduffee, one way to make it work would be to tell LXD to ignore the ZFS kernel module version it detects and blindly use ZFS 2.1 tooling with it. ZFS tools support interacting with different kernel module versions but it's not ideal/officially supported so your mileage may vary. It's a hack.

Please make sure to be using 18.04's [HWE kernel](https://ubuntu.com/kernel/lifecycle#installation-18-04) that comes with ZFS 0.8 as my tests with the GA kernel (ZFS 0.7) didn't work.

Here's how to do this:

```
# uname -r | grep ^5\.4
5.4.0-150-generic
# cp /snap/lxd/current/commands/daemon.start .
# vim daemon.start
```

When editing the `daemon.start` file, add the following:

```
$ git diff
diff --git a/snapcraft/commands/daemon.start b/snapcraft/commands/daemon.start
index 79ed2747..e3359304 100755
--- a/snapcraft/commands/daemon.start
+++ b/snapcraft/commands/daemon.start
@@ -353,20 +353,24 @@ else
 fi
 
 ZFS_VER="$(echo "${VERSION}" | cut -c 1-3)"
 
 if [ -d "${SNAP_CURRENT}/zfs-${ZFS_VER}/bin" ]; then
     echo "==> Setting up ZFS (${ZFS_VER})"
     export LD_LIBRARY_PATH="${SNAP_CURRENT}/zfs-${ZFS_VER}/lib/:${LD_LIBRARY_PATH}"
     export PATH="${SNAP_CURRENT}/zfs-${ZFS_VER}/bin:${PATH}"
 elif [ -n "${ZFS_VER}" ]; then
     echo "==> Unsupported ZFS version (${ZFS_VER})"
+    ZFS_VER="$(find "${SNAP_CURRENT}/" -type d -name 'zfs-*' | cut -d- -f2| sort -n | head -n1)"
+    echo "==> Falling back to ZFS version ${ZFS_VER}"
+    export LD_LIBRARY_PATH="${SNAP_CURRENT}/zfs-${ZFS_VER}/lib/:${LD_LIBRARY_PATH}"
+    export PATH="${SNAP_CURRENT}/zfs-${ZFS_VER}/bin:${PATH}"
 else
     echo "==> No ZFS support"
 fi
 
 # Escape resource limits
 echo "==> Escaping the systemd cgroups"
 if [ -e "/sys/fs/cgroup/cgroup.procs" ]; then
     echo "====> Detected cgroup V2"
 
     # Try to escape to the root (V2 hosts)
```

Then apply the changes:

```
# mount -o ro,bind daemon.start /snap/lxd/current/commands/daemon.start
# snap restart lxd
```

**Comment by amcduffee** [comment]
> If your using a cluster also add `--cohort=”+”` see
> 
> https://documentation.ubuntu.com/lxd/en/latest/howto/snap/#keep-cluster-members-in-sync

I am trying to understand how the cohort feature works.  Does the `+` have a special meaning?  I also see the `--leave-cohort` option and it makes me wonder if you are suggesting that I use cohorts to safely get all cluster members over onto a different tracking branch, after which I have them all leave the cohort?  What are the differences and risks in just running the above refresh commands without using cohorts?

**Comment by simondeziel** [comment]
The `+` has special meaning that essentially boils down to "give me the latest snap revision available, ignore any phasing". This way, all the cluster nodes will see the latest revision rather they having some of them see an older revision due to the phasing mechanism.

**Comment by amcduffee** [comment]
> @amcduffee, one way to make it work would be to tell LXD to ignore the ZFS kernel module version it detects and blindly use ZFS 2.1 tooling with it. ZFS tools support interacting with different kernel module versions but it's not ideal/officially supported so your mileage may vary. It's a hack.
> 
> Please make sure to be using 18.04's [HWE kernel](https://ubuntu.com/kernel/lifecycle#installation-18-04) that comes with ZFS 0.8 as my tests with the GA kernel (ZFS 0.7) didn't work.
> 
> Here's how to do this:
> 
> ```
> # uname -r | grep ^5\.4
> 5.4.0-150-generic
> # cp /snap/lxd/current/commands/daemon.start .
> # vim daemon.start
> ```
> 
> When editing the `daemon.start` file, add the following:
> 
> ```
> $ git diff
> diff --git a/snapcraft/commands/daemon.start b/snapcraft/commands/daemon.start
> index 79ed2747..e3359304 100755
> --- a/snapcraft/commands/daemon.start
> +++ b/snapcraft/commands/daemon.start
> @@ -353,20 +353,24 @@ else
>  fi
>  
>  ZFS_VER="$(echo "${VERSION}" | cut -c 1-3)"
>  
>  if [ -d "${SNAP_CURRENT}/zfs-${ZFS_VER}/bin" ]; then
>      echo "==> Setting up ZFS (${ZFS_VER})"
>      export LD_LIBRARY_PATH="${SNAP_CURRENT}/zfs-${ZFS_VER}/lib/:${LD_LIBRARY_PATH}"
>      export PATH="${SNAP_CURRENT}/zfs-${ZFS_VER}/bin:${PATH}"
>  elif [ -n "${ZFS_VER}" ]; then
>      echo "==> Unsupported ZFS version (${ZFS_VER})"
> +    ZFS_VER="$(find "${SNAP_CURRENT}/" -type d -name 'zfs-*' | cut -d- -f2| sort -n | head -n1)"
> +    echo "==> Falling back to ZFS version ${ZFS_VER}"
> +    export LD_LIBRARY_PATH="${SNAP_CURRENT}/zfs-${ZFS_VER}/lib/:${LD_LIBRARY_PATH}"
> +    export PATH="${SNAP_CURRENT}/zfs-${ZFS_VER}/bin:${PATH}"
>  else
>      echo "==> No ZFS support"
>  fi
>  
>  # Escape resource limits
>  echo "==> Escaping the systemd cgroups"
>  if [ -e "/sys/fs/cgroup/cgroup.procs" ]; then
>      echo "====> Detected cgroup V2"
>  
>      # Try to escape to the root (V2 hosts)
> ```
> 
> Then apply the changes:
> 
> ```
> # mount -o ro,bind daemon.start /snap/lxd/current/commands/daemon.start
> # snap restart lxd
> ```

Thanks for the idea, if nothing else it is interesting to know this can be done.

Truthfully, I need to get the machine updated to a newer LTS, but there hasn't been anything to force it to happen, until now.

**Comment by amcduffee** [comment]
> The `+` has special meaning that essentially boils down to "give me the latest snap revision available, ignore any phasing". This way, all the cluster nodes will see the latest revision rather they having some of them see an older revision due to the phasing mechanism.

Is the phasing a feature on the snap store servers?  To stagger (or phase) updates rolling out to clients?

I can see all of my nodes are `in-cohort` already:
```
installed:          5.21.1-43998c6            (28155) 107MB in-cohort
```
Is there a way to see what cohort they are in?

**Comment by amcduffee** [comment]
I found the current cohort value by using `snap info --verbose lxd`.  I don't recall ever setting the cohort.  Is that something LXD did for me when I built the cluster?

**Comment by tomponline** [comment]
@amcduffee please take a look at https://discourse.ubuntu.com/t/lxd-5-21-1-lts-has-been-released/43823/24 it may allow you to get back to lxd 5.20 and then use snap hold to pin to it.

**Comment by justinclift** [comment]
> Yes it was in the release notes about the change in requirement and the release was in the candidate channel for several weeks.

Bear in mind that this has been pushed out to LTS distributions, whose whole purpose in life is to NOT spontaneously break. :frowning: :frowning: :frowning:

**Comment by justinclift** [comment]
> We are now aware in particular of users on Ubuntu 18.04 ...

For us, it was two production servers on Ubuntu 20.04 (without HWE kernel) set up a few years ago.  It does look like the snap version of LXD originally set up kept switching to whatever the newest channel was.

For us, both of these servers *did* have the older version of the lxd database (`global.bak`) still in place, so could be rolled back.  We've naturally now frozen the version of lxd on those since rolling back, which is nice to see can be done these days.  Older versions of snap wouldn't allow for holding back updates more than 30 days (!!!).

Going forward though, we've move most other stuff to Debian 12.  Ubuntu is no longer an option.  Not due just to this, but also due to pushing snap itself despite Community sentiment + also pushing advertising ***in the damn cli***.

Thanks for your efforts in figuring out a solution for the remaining Ubuntu holdouts @tomponline. :smile:

**Comment by tomponline** [comment]
Just to update you all on the ZFS on Bionic and Focal situation.

We’ve decided to reinstate ZFS 0.8 userland tooling into the 5.21/* channels only (but not latest/stable).

This will give those tracking latest/stable on Bionic and Focal a chance to switch to the 5.21/stable LTS series.

Ubuntu Bionic users will still need to switch to the HWE kernel to get ZFS 0.8 support.
Ubuntu Focal users will be able to use either standard or HWE kernels.

This is effectively a time-limited solution as the latest/stable channel will continue to move on with each feature release in the coming months and may introduce further DB schema changes which will then prevent switching to 5.21/stable.

But for now latest/stable and 5.21/stable are equivalent and so switching is possible.

The change is going into 5.21/stable now.

https://github.com/canonical/lxd-pkg-snap/pull/421

**Comment by xrobau** [comment]
If I may stick my head in here as someone who was affected that purely coincidentally happened to be the guy running Linux.conf.au last week (now called Everything Open), I mentioned this as an example of a terrible UX that could have been solved in a much less catastrophic way by splitting these two major changes into two seperate releases.

If you had pushed a release out that *only* updated the required ZFS version, everything would have broken and people would have rolled back to the previous snap, pinned it, and the database would have been fine. 

If you had pushed out a release that *only* updated the schema, nothing would have broken until the ZFS upgrade, people would have rolled back to the previous snap, pinned it, and everything STILL would have been fine.

Regardless of LTS or non-LTS, doing two major things like that in the same release is something that should be discouraged in any project.  I bring this up purely as an open source developer and User Experience Advocate, and I hope that this (relatively minor in the grand scheme of things) incident can stick in people's minds in the future.

**Comment by tomponline** [comment]
@sunebeck what kernel & ubuntu version are you running and what does `snap info lxd` show?

You maybe able to get up and running again more easily by using an HWE kernel on `5.21/stable` channel.

See https://discourse.ubuntu.com/t/lxd-5-21-1-lts-has-been-released/43823

**Comment by justinclift** [comment]
@sunebeck I followed the steps that @tomponline is linking to in the comment just above this one I'm writing now.

The link url `...lxd-5-21-1-lts-has-been-released...` makes it looks like just a release announcement post, but it also has the step-by-step instructions for rolling back to the previous version if there's still a .bak folder in the right spot.

**Comment by justinclift** [comment]
@tomponline Might be a better idea to use a major version bump (lxd 6!) next time there's a significant breaking change too. :wink:

**Comment by tomponline** [comment]
@justinclift LXD 5.21.x was a major version bump as well as the start of a new LTS series, see https://discourse.ubuntu.com/t/lxd-5-21-0-lts-has-been-released/42476

The reason most folks are seeing this on LXD 5.21.1 is because this was the first release in that series to make it to the stable channels.

We don't change supported ZFS versions during an LTS series and that hasn't happened in this case.

**Comment by tomponline** [comment]
> The link url `...lxd-5-21-1-lts-has-been-released...` makes it looks like just a release announcement post, but it also has the step-by-step instructions for rolling back to the previous version if there's still a .bak folder in the right spot.

I updated it with the anchor I intended it to go to. Discourse seems to remove the anchor when you click it confusingly.

**Comment by sunebeck** [comment]
> We've included ZFS 0.8 in the 5.21/stable channel now, so if you switch to that you can avoid needing to revert and pin to an unsupported feature channel (e.g. like 5.20/stable).

That seems easier, thanks! Oddly enough, now getting a database error instead after doing that:

> time="2024-04-25T15:56:21Z" level=error msg="Failed to start the daemon" err="Failed to start dqlite server: raft_start(): io: load open segment open-1: metadata contains term 1"

On Debian 12, lxd `5.21.1-98dad8f`. Might be a different issue, but perhaps still related to the upgrade?

**Comment by tomponline** [comment]
> On Debian 12, lxd `5.21.1-98dad8f`

I'm surprised you had issues with ZFS 0.8 removal if running on Debian 12, as it appears to come with ZFS 2.1.

**Comment by sunebeck** [comment]
> I'm surprised you had issues with ZFS 0.8 removal if running on Debian 12, as it appears to come with ZFS 2.1.

Oops, my bad, it's indeed Debian 11. Sorry!

**Comment by cole-miller** [comment]
Hi @sunebeck, is this error repeatable? What's the output of `sudo ls -lah /var/snap/lxd/common/lxd/database/global`? Can you follow [these instructions](https://discourse.ubuntu.com/t/lxd-5-21-1-lts-has-been-released/43823/67) to get tracing output? Thanks!

**Comment by christianflintrup** [comment]
@cole-miller After restarting lxd once more to see what happened, a new error occurred:

```
time="2024-04-25T16:45:38Z" level=error msg="Failed to start the daemon" err="Failed to start dqlite server: raft_start(): io: closed segment 0000000000049266-0000000000049266 is past last snapshot snapshot-1-49152-31849044399"
```

I suspect I accidentally may have run the `lxd` daemon twice simultaneously; I typed `lxd` instead of `lxc` by accident, while snap might also have been in the process of restarting it.

For some reason, `common/lxd/logs/lxd.log` is log-rotated very aggressively, each file only containing a few lines.

Snap info:

```
tracking:     5.21/stable
refresh-date: today at 13:18 UTC
...
installed:          5.21.1-98dad8f            (28323) 108MB -
```

Here's the dqlite diretory:

```
total 89M
drwxr-x--- 2 root root  4.0K Apr 25 16:50 .
drwx------ 5 root root  4.0K Apr 25 15:56 ..
-rw------- 1 root root  8.0M Feb 21 07:12 0000000000040961-0000000000041735
-rw------- 1 root root  2.6M Feb 23 10:12 0000000000041736-0000000000041984
-rw------- 1 root root  8.0M Mar  1 09:12 0000000000041985-0000000000042762
-rw------- 1 root root  2.6M Mar  3 08:12 0000000000042763-0000000000043008
-rw------- 1 root root  8.0M Mar  9 23:12 0000000000043009-0000000000043793
-rw------- 1 root root  2.7M Mar 11 16:12 0000000000043794-0000000000044032
-rw------- 1 root root  8.0M Mar 18 07:12 0000000000044033-0000000000044809
-rw------- 1 root root  2.6M Mar 20 08:12 0000000000044810-0000000000045056
-rw------- 1 root root  8.0M Mar 27 07:12 0000000000045057-0000000000045836
-rw------- 1 root root  2.6M Mar 29 11:12 0000000000045837-0000000000046080
-rw------- 1 root root  8.0M Apr  5 02:12 0000000000046081-0000000000046861
-rw------- 1 root root  2.5M Apr  7 06:12 0000000000046862-0000000000047104
-rw------- 1 root root  8.0M Apr 14 02:12 0000000000047105-0000000000047881
-rw------- 1 root root  2.6M Apr 16 06:12 0000000000047882-0000000000048128
-rw------- 1 root root  8.0M Apr 23 01:12 0000000000048129-0000000000048904
-rw------- 1 root root  2.1M Apr 24 16:37 0000000000048905-0000000000049098
-rw------- 1 root root  762K Apr 24 16:37 0000000000049099-0000000000049152
-rw------- 1 root root  1.2M Apr 24 16:37 0000000000049153-0000000000049251
-rw------- 1 root root   17K Apr 24 16:37 0000000000049252-0000000000049252
-rw------- 1 root root   17K Apr 24 16:37 0000000000049253-0000000000049253
-rw------- 1 root root   17K Apr 24 16:37 0000000000049254-0000000000049254
-rw------- 1 root root   17K Apr 24 16:37 0000000000049255-0000000000049255
-rw------- 1 root root   17K Apr 24 16:37 0000000000049256-0000000000049256
-rw------- 1 root root   17K Apr 25 13:18 0000000000049257-0000000000049257
-rw------- 1 root root   17K Apr 25 13:18 0000000000049258-0000000000049258
-rw------- 1 root root   17K Apr 25 13:18 0000000000049259-0000000000049259
-rw------- 1 root root   17K Apr 25 13:18 0000000000049260-0000000000049260
-rw------- 1 root root   17K Apr 25 13:18 0000000000049261-0000000000049261
-rw------- 1 root root   17K Apr 25 13:18 0000000000049262-0000000000049262
-rw------- 1 root root   17K Apr 25 15:56 0000000000049263-0000000000049263
-rw------- 1 root root   17K Apr 25 15:56 0000000000049264-0000000000049264
-rw------- 1 root root   17K Apr 25 15:56 0000000000049265-0000000000049265
-rw------- 1 root root   17K Apr 25 15:56 0000000000049266-0000000000049266
-rw------- 1 root root  125K Apr 25 15:56 0000000000049266-0000000000049275
-rw-r--r-- 1 root root 1004K Apr 25 15:56 db.bin
-rw------- 1 root root  1.1M Apr 25 15:56 db.bin-wal
-rw------- 1 root root    32 Apr 22  2023 metadata1
-rw------- 1 root root  167K Apr 16 06:12 snapshot-1-48128-31120326220
-rw------- 1 root root    56 Apr 16 06:12 snapshot-1-48128-31120326220.meta
-rw------- 1 root root  267K Apr 24 16:37 snapshot-1-49152-31849044399
-rw------- 1 root root    56 Apr 24 16:37 snapshot-1-49152-31849044399.meta
```

P.S: Tried making a post on discourse instead of in this thread, but got "Access Denied" — maybe because the user was just created?

**Comment by sunebeck** [comment]
@cole-miller Actually a different error shows up after a second restart:

```
Failed to start dqlite server: raft_start(): io: closed segment 0000000000049266-0000000000049266 is past last snapshot snapshot-1-49152-31849044399"
```

Here's the database directory:

```
# ls -lah /var/snap/lxd/common/lxd/database/global
total 89M
drwxr-x--- 2 root root  4.0K Apr 25 19:51 .
drwx------ 5 root root  4.0K Apr 25 15:56 ..
-rw------- 1 root root  8.0M Feb 21 07:12 0000000000040961-0000000000041735
-rw------- 1 root root  2.6M Feb 23 10:12 0000000000041736-0000000000041984
-rw------- 1 root root  8.0M Mar  1 09:12 0000000000041985-0000000000042762
-rw------- 1 root root  2.6M Mar  3 08:12 0000000000042763-0000000000043008
-rw------- 1 root root  8.0M Mar  9 23:12 0000000000043009-0000000000043793
-rw------- 1 root root  2.7M Mar 11 16:12 0000000000043794-0000000000044032
-rw------- 1 root root  8.0M Mar 18 07:12 0000000000044033-0000000000044809
-rw------- 1 root root  2.6M Mar 20 08:12 0000000000044810-0000000000045056
-rw------- 1 root root  8.0M Mar 27 07:12 0000000000045057-0000000000045836
-rw------- 1 root root  2.6M Mar 29 11:12 0000000000045837-0000000000046080
-rw------- 1 root root  8.0M Apr  5 02:12 0000000000046081-0000000000046861
-rw------- 1 root root  2.5M Apr  7 06:12 0000000000046862-0000000000047104
-rw------- 1 root root  8.0M Apr 14 02:12 0000000000047105-0000000000047881
-rw------- 1 root root  2.6M Apr 16 06:12 0000000000047882-0000000000048128
-rw------- 1 root root  8.0M Apr 23 01:12 0000000000048129-0000000000048904
-rw------- 1 root root  2.1M Apr 24 16:37 0000000000048905-0000000000049098
-rw------- 1 root root  762K Apr 24 16:37 0000000000049099-0000000000049152
-rw------- 1 root root  1.2M Apr 24 16:37 0000000000049153-0000000000049251
-rw------- 1 root root   17K Apr 24 16:37 0000000000049252-0000000000049252
-rw------- 1 root root   17K Apr 24 16:37 0000000000049253-0000000000049253
-rw------- 1 root root   17K Apr 24 16:37 0000000000049254-0000000000049254
-rw------- 1 root root   17K Apr 24 16:37 0000000000049255-0000000000049255
-rw------- 1 root root   17K Apr 24 16:37 0000000000049256-0000000000049256
-rw------- 1 root root   17K Apr 25 13:18 0000000000049257-0000000000049257
-rw------- 1 root root   17K Apr 25 13:18 0000000000049258-0000000000049258
-rw------- 1 root root   17K Apr 25 13:18 0000000000049259-0000000000049259
-rw------- 1 root root   17K Apr 25 13:18 0000000000049260-0000000000049260
-rw------- 1 root root   17K Apr 25 13:18 0000000000049261-0000000000049261
-rw------- 1 root root   17K Apr 25 13:18 0000000000049262-0000000000049262
-rw------- 1 root root   17K Apr 25 15:56 0000000000049263-0000000000049263
-rw------- 1 root root   17K Apr 25 15:56 0000000000049264-0000000000049264
-rw------- 1 root root   17K Apr 25 15:56 0000000000049265-0000000000049265
-rw------- 1 root root   17K Apr 25 15:56 0000000000049266-0000000000049266
-rw------- 1 root root  125K Apr 25 15:56 0000000000049266-0000000000049275
-rw-r--r-- 1 root root 1004K Apr 25 15:56 db.bin
-rw------- 1 root root  1.1M Apr 25 15:56 db.bin-wal
-rw------- 1 root root    32 Apr 22  2023 metadata1
-rw------- 1 root root  167K Apr 16 06:12 snapshot-1-48128-31120326220
-rw------- 1 root root    56 Apr 16 06:12 snapshot-1-48128-31120326220.meta
-rw------- 1 root root  267K Apr 24 16:37 snapshot-1-49152-31849044399
-rw------- 1 root root    56 Apr 24 16:37 snapshot-1-49152-31849044399.meta
```

I suspect I might have run `lxd` twice simultaneously, because I typed `lxd` instead of `lxc` at one point. Do you think that could have any relevance? 

P.S: Tried creating a thread on discourse instead, since it seems less relevant to this topic now. But it gave me an “Access Denied” error.

**Comment by tomponline** [comment]
> `Failed to start dqlite server: raft_start(): io: closed segment 0000000000049266-0000000000049266 is past last snapshot snapshot-1-49152-31849044399"`

The docs @cole-miller is writing here https://github.com/canonical/lxd/pull/13346 should help.

**Comment by cole-miller** [comment]
Thanks for the information @sunebeck. I'd suggest:

1. back up that directory in its entirety
2. trying removing `0000000000049266-0000000000049266` from the directory and starting LXD (and, if it starts successfully, confirming that it's in the shape you expect)
3. if that fails, the same with `0000000000049266-0000000000049275`

You'll notice that the (inclusive) ranges represented by those two filenames *overlap*; dqlite tries to maintain an invariant that this never happens, but in this case something went wrong, perhaps related to running two dqlites concurrently with the same directory as you mentioned.

**Comment by sunebeck** [comment]
> What is your username?

My discourse username is `sunebeck`. I just created it, so maybe there's some anti-spam measure in effect?

> You'll notice that the (inclusive) ranges represented by those two filenames overlap

Oh, right! I actually did check your new sqlite documentation yesterday, and checked the ranges, but just somehow failed at noticing the overlap.

Removing `0000000000049266-0000000000049266` helped and everything is back to normal! Thanks, and sorry for hijacking this issue.
